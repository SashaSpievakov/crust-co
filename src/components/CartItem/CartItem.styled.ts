import styled from 'styled-components';

const Article = styled.article`
  display: flex;
  width: 100%;
  position: relative;
  border-top: 1px solid ${({ theme }) => theme.colors.gray};
  padding-top: 30px;
  margin-top: 30px;

  @media screen and (max-width: 850px) {
    flex-direction: column;
    row-gap: 10px;
    align-items: center;
  }
`;

const ImgWrapper = styled.div`
  display: flex;
  align-items: center;
  margin-right: 15px;
  width: 10%;

  img {
    border-radius: 20px;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    margin-right: 0;

    img {
      width: 200px;
      height: 200px;
    }
  }
`;

const InfoWrapper = styled.section`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 40%;

  h3 {
    font-weight: bold;
    font-size: 22px;
    line-height: 27px;
    letter-spacing: 0.01em;
  }

  p {
    font-size: 18px;
    color: #8d8d8d;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    text-align: center;
  }
`;

const Counter = styled.section`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 13%;

  span {
    font-size: 20px;
  }

  @media screen and (max-width: 850px) {
    width: auto;
    gap: 11px;
  }
`;

const CountIcon = styled.i`
  stroke-width: 1;
`;

const Price = styled.h3`
  font-size: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 33%;
`;

const Remove = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  width: 4%;

  @media screen and (max-width: 850px) {
    position: absolute;
    right: 0;
    top: 20px;
  }
`;

export { Article, Counter, CountIcon, ImgWrapper, InfoWrapper, Price, Remove };
