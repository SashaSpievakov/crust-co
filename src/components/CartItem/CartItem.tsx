import { memo } from 'react';
import { BsDashLg, BsPlusLg } from 'react-icons/bs';

import { addItem, removeItem, removeItems } from '../../store/slices/cartSlice';
import { ButtonCircle } from '../../styles/Buttons.styled';
import {
  Section,
  ImgWrapper,
  InfoWrapper,
  Counter,
  CountIcon,
  Price,
  Remove,
} from './CartItem.styled';
import { ICartItem } from '../../models/ICartItem';
import { useAppDispatch } from '../../hooks/reduxHooks';

const CartItem = memo(({ id, name, price, size, type, count }: ICartItem) => {
  const dispatch = useAppDispatch();

  const onClickRemove = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(removeItem(item));
  };

  const onClickAdd = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(addItem(item));
  };

  const onClickDelete = () => {
    const item: ICartItem = {
      id,
      name,
      price,
      size,
      type,
      count,
    };
    dispatch(removeItems(item));
  };

  return (
    <Section>
      <ImgWrapper>
        <img
          src={`./assets/img/pizza${id}.png`}
          alt="Pizza"
          width="80"
          height="80"
        />
      </ImgWrapper>
      <InfoWrapper>
        <h3>{name}</h3>
        <p>
          {type} dough, {size} inch
        </p>
      </InfoWrapper>
      <Counter>
        <ButtonCircle
          disabled={count === 1}
          onClick={onClickRemove}
          aria-label="minus"
        >
          <CountIcon as={BsDashLg} />
        </ButtonCircle>
        <span>{count}</span>
        <ButtonCircle onClick={onClickAdd} aria-label="plus">
          <CountIcon as={BsPlusLg} />
        </ButtonCircle>
      </Counter>
      <Price>
        <b>{price * count}$</b>
      </Price>
      <Remove>
        <ButtonCircle onClick={onClickDelete} remove aria-label="delete">
          <CountIcon as={BsPlusLg} />
        </ButtonCircle>
      </Remove>
    </Section>
  );
});
export default CartItem;
