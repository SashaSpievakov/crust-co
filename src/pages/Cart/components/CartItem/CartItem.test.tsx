import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '@src/tests/helpers';
import {
  CartItemMockProps,
  CartItemMockProps2,
  CartItemsMockProps,
} from '@src/tests/mocks/mockData/mockData';

import { CartItem } from './CartItem';

describe('CartItem tests', () => {
  test('should render the CartItem component', () => {
    const snapshot = rendererWithProviders(<CartItem {...CartItemMockProps} />);
    expect(snapshot).toMatchSnapshot();
  });

  test('should render the CartItem component with a disabled button', () => {
    const snapshot = rendererWithProviders(
      <CartItem {...CartItemMockProps2} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('CartItem button tests', () => {
    beforeEach(() => {
      renderWithProvidersAndRoutes(null, true, '/cart', {
        cart: CartItemsMockProps,
      });
    });

    test('should remove an item', () => {
      const buttons = screen.getAllByRole('button');
      const count = screen.getByText(/13/i);
      const price = screen.getByText(/91/i);

      userEvent.click(buttons[0]);
      userEvent.click(buttons[0]);

      expect(count).toHaveTextContent('11');
      expect(price).toHaveTextContent('77');
    });

    test('should add an item', () => {
      const buttons = screen.getAllByRole('button');
      const count = screen.getByText(/13/i);
      const price = screen.getByText(/91/i);

      userEvent.click(buttons[1]);
      userEvent.click(buttons[1]);

      expect(count).toHaveTextContent('15');
      expect(price).toHaveTextContent('105');
    });

    test('should delete all item', () => {
      const buttons = screen.getAllByRole('button');

      userEvent.click(buttons[2]);
      userEvent.click(buttons[5]);

      expect(
        screen.getByRole('heading', {
          name: /your cart is empty/i,
        }),
      ).toBeInTheDocument();
    });

    test('should change global count', () => {
      const buttons = screen.getAllByRole('button');
      const count = screen.getByTestId('cartPageItemsCount');

      userEvent.click(buttons[0]);
      userEvent.click(buttons[1]);
      userEvent.click(buttons[1]);
      userEvent.click(buttons[4]);
      userEvent.click(buttons[4]);

      expect(count).toHaveTextContent('17');
    });

    test('should change global price', () => {
      const buttons = screen.getAllByRole('button');
      const count = screen.getByTestId('cartPageItemsPrice');

      userEvent.click(buttons[0]);
      userEvent.click(buttons[1]);
      userEvent.click(buttons[1]);
      userEvent.click(buttons[4]);
      userEvent.click(buttons[4]);

      expect(count).toHaveTextContent('125');
    });
  });
});
