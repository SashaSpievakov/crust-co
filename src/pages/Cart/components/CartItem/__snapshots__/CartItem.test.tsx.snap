// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem tests should render the CartItem component 1`] = `
Object {
  "children": Array [
    <div
      className="sc-jfvxQR kowwjN"
    >
      <picture>
        <source
          srcSet="./assets/img/pizzas/pizza6.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizzas/pizza6.png"
          type="image/png"
        />
        <img
          alt="Cheesburger Pizza pizza"
          height="80"
          src="./assets/img/pizzas/pizza6.png"
          width="80"
        />
      </picture>
    </div>,
    <div
      className="sc-eJDSGI kzQfpt"
    >
      <a
        href="/pizzas/Cheesburger Pizza"
        onClick={[Function]}
      >
        Cheesburger Pizza
      </a>
      <p>
        thin
         dough, 
        14
         inch
      </p>
    </div>,
    <section
      className="sc-oZIhv kHGFxV"
    >
      <button
        aria-label="minus"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF ggHeaE"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
          />
        </svg>
      </button>
      <span>
        13
      </span>
      <button
        aria-label="plus"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF ggHeaE"
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </section>,
    <h3
      className="sc-ehvNnt jGsMfb"
    >
      91
      $
    </h3>,
    <div
      className="sc-laZRCg eTmGYj"
    >
      <button
        aria-label="delete"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF gjJRSz"
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </div>,
  ],
  "props": Object {
    "className": "sc-iOeugr bMiyJG",
  },
  "type": "article",
}
`;

exports[`CartItem tests should render the CartItem component with a disabled button 1`] = `
Object {
  "children": Array [
    <div
      className="sc-jfvxQR kowwjN"
    >
      <picture>
        <source
          srcSet="./assets/img/pizzas/pizza2.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizzas/pizza2.png"
          type="image/png"
        />
        <img
          alt="Pepperoni Pizza pizza"
          height="80"
          src="./assets/img/pizzas/pizza2.png"
          width="80"
        />
      </picture>
    </div>,
    <div
      className="sc-eJDSGI kzQfpt"
    >
      <a
        href="/pizzas/Pepperoni Pizza"
        onClick={[Function]}
      >
        Pepperoni Pizza
      </a>
      <p>
        traditional
         dough, 
        16
         inch
      </p>
    </div>,
    <section
      className="sc-oZIhv kHGFxV"
    >
      <button
        aria-label="minus"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF ggHeaE"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
          />
        </svg>
      </button>
      <span>
        1
      </span>
      <button
        aria-label="plus"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF ggHeaE"
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </section>,
    <h3
      className="sc-ehvNnt jGsMfb"
    >
      9
      $
    </h3>,
    <div
      className="sc-laZRCg eTmGYj"
    >
      <button
        aria-label="delete"
        className="sc-bcXHqe sc-eDvSVe sc-jSUZER glVyWT fwWQTF gjJRSz"
        onClick={[Function]}
      >
        <svg
          className="sc-hiDMwi fNuffT"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
      </button>
    </div>,
  ],
  "props": Object {
    "className": "sc-iOeugr bMiyJG",
  },
  "type": "article",
}
`;
